snap/info.go

```diff
@@ -1014,11 +1021,20 @@ func (s *Info) MangleDesktopFileName(desktopFile string) (string, error) {
 			return filepath.Join(dir, base), nil
 		}
 	}
-	// FIXME: don't blindly use the snap desktop filename, mangle it
-	// but we can't just use the app name because a desktop file
-	// may call the same app with multiple parameters, e.g.
-	// --create-new, --open-existing etc
-	return filepath.Join(dir, fmt.Sprintf("%s_%s", s.DesktopPrefix(), base)), nil
+
+	// Sanitization logic shouldn't worry about being backware compatible because the
+	// desktop files are always written when snapd starts in ensureDesktopFilesUpdated.
+	sanitizedBase := ""
+	for _, c := range base {
+		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_' || c == '-' || c == '.' {
+			sanitizedBase += string(c)
+			continue
+		}
+		// Replace with '_'
+		sanitizedBase += "_"
+	}
```

Comment:
I think I'd split this out into a separate function. Instead of concatenating string, I'd suggest to use a string builder:
```
var b strings.Builder
b.Grow(len(base)
for _, c := range base {
    if ... {
        b.WriteRune(c)
    } else {
        b.WriteRune('_')
    }
}
return b.String()
```